### PR TITLE
libtorrent 1.2 + boost 1.70 or libtorrent 2.0 + boost 1.78

### PR DESCRIPTION
**Pull request by th0ma7**
## Description

Demonstrate feasibility to build build libtorrent using crossenv, part 2.  The following uses only supported version:
- [x] `libtorrent` 2.0.5 (using `boost` 1.78) - DSM7 except `comcerto2k-7.0`
- [x] `libtorrent` 1.2.16 (using `boost` 1.70) - DSM6 + DSM7
- [x] `rencode` 1.0.7 (from a fork of the project as no-longer maintained)
- [x] `deluge` 2.0.5

Follows-up #5180
Parallel work to #4153

`libtorrent` bug with DSM6/comcerto2k-7.0 using gcc-4.9.x compiler: https://github.com/arvidn/libtorrent/issues/6868

## Checklist

- [ ] Build rule `all-supported` completed successfully
- [ ] New installation of package completed successfully
- [ ] Package upgrade completed successfully (Manually install the package again)
- [ ] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Update-Policy#tests-checks)
- [ ] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created

### Type of change

- [ ] Bug fix
- [ ] New Package
- [x] Package update
- [x] Includes small framework changes
- [ ] This change requires a documentation update (e.g. Wiki)

### CODE REVIEW

File: mk/spksrc.common.mk [hunk 1]
```diff
@@ -119,3 +119,6 @@ version_le = $(shell if printf '%s\n' "$(1)" "$(2)" | sort -VC ; then echo 1; fi
 version_ge = $(shell if printf '%s\n' "$(1)" "$(2)" | sort -VCr ; then echo 1; fi)
 version_lt = $(shell if [ "$(1)" != "$(2)" ] && printf "%s\n" "$(1)" "$(2)" | sort -VC ; then echo 1; fi)
 version_gt = $(shell if [ "$(1)" != "$(2)" ] && printf "%s\n" "$(1)" "$(2)" | sort -VCr ; then echo 1; fi)
+
+# GCC version
+version_gcc = $(shell $$(sed -n -e '/TC_ENV += TC_GCC/ s/.*= *//p' $(PWD)/$(subst arch,work,$(MAKECMDGOALS))/tc_vars.mk))
```
Comment (th0ma7) [inline mk/spksrc.common.mk:124]: @hgy59 I'd like to bring your attention to this function helper so we can finally make usage of the GCC version from any spk.  See use-case in python310 Makefile which allows to capture `comcerto2k-7.0` exception cleanly.

**Comment by th0ma7** [comment]
@hgy59 I think I'm up to a working solution using `libtorrent` 1.2.16 + `boost` 1.70 which I got to build properly using **crossenv** on both DSM6+7 with exception of ARMv5 and ARMv7L due to GCC being too old (GCC < 4.9, even using `c++1x`).

I may even be inches away from a success with `libtorrent` 2.0.6 + `boost` 1.76 but I'm lacking a fix for using `c++1y` with GCC < 4.9 (bug opened upstream https://github.com/arvidn/libtorrent/issues/6868).

Next would be to move the code out from `python310` and back to `deluge` and initiate actual testing.

File: mk/spksrc.common.mk (hunk 1 above)
Comment (hgy59) [reply mk/spksrc.common.mk]: My findings so far:
- you don't have `TC_ENV` nor `TC_GCC` available before the target `$(TC_VARS_MK)` is built.
- you can modify makefile variables in a custom `PRE_DEPEND_TARGET` that depends on `$(TC_VARS_MK)` with $(eval )

``` 
.PHONY: python310_pre_depend
python310_pre_depend: $(TC_VARS_MK)
	@$(MSG) "Evaluate dependencies for GCC version $(TC_GCC)"
	$(eval WHEELS_BUILD_ARGS += $(shell if "$(TC_GCC)" != "5.0" && printf '%s\n' "$(TC_GCC)" "5.0" | sort -VC ; then echo --b2-args=cxxstd=11 ; fi))
	@$(MSG) "WHEELS_BUILD_ARGS=$(WHEELS_BUILD_ARGS)"
``` 

- ensure the `$(TC_VARS_MK)` is built before
- version comparision migth be improved (but `version_lt` is not ment for shell environment)

**Comment by hgy59** [comment]
@th0ma7 we probably sould first merge this for libtorrent 1.2 to create packages for DSM 6 and DSM 7?

**Comment by th0ma7** [comment]
> @th0ma7 we probably sould first merge this for libtorrent 1.2 to create packages for DSM 6 and DSM 7?

It's exacltly where I was heading this morning.  I was planning to clean-up slightly the code and recomment python310 test/demo wheels and then switch to @Strangelovian branch to include the various bits needed.

File: mk/spksrc.common.mk (hunk 1 above)
Comment (th0ma7) [reply mk/spksrc.common.mk]: Your findings where pretty much the route I was originally leaning towards but the relatively simplistic method I've moved to uses the `TC_GCC` version variable from the `tc_vars.mk` and works quite well.

It actually solves a long-standing issue I had with:
```
ifeq ($(call version_ge, ${TCVERSION}, 7.0),1)
# fix compilation with newer compilers
ADDITIONAL_CXXFLAGS += -D_GLIBCXX_USE_C99_MATH -D_GLIBCXX_HAVE_OBSOLETE_ISNAN -D_GLIBCXX_HAVE_OBSOLETE_ISINF
endif
```
That can then be converted into (and takes into consideration `comcerto2k-7.0`):
```
ifeq ($(call version_ge, $(call version_gcc), 7.5),1)
# fix compilation with newer compilers
ADDITIONAL_CXXFLAGS += -D_GLIBCXX_USE_C99_MATH -D_GLIBCXX_HAVE_OBSOLETE_ISNAN -D_GLIBCXX_HAVE_OBSOLETE_ISINF
endif
```

**Comment by th0ma7** [comment]
@hgy59 if you're OK with the proposed changes I'd move ahead and merge.  I already have changes pending for  @Strangelovian PR to re-adjust accordingly.

**Review by hgy59 — APPROVED** [review]
if version _gcc really works, we should introduce REQUIRED_MIN_GCC some time.
So far it looks good to me for merging.

**Comment by th0ma7** [comment]
> if version _gcc really works, we should introduce REQUIRED_MIN_GCC some time. 

That's an awesome idea :)

**Pull request merged**